[NativeAOT] Add ability to generate library and exe entry points

AustinWise opened the pull request:
This PR adds a new flag to ILC, `--split-exe-initalization` . The flag splits the initialization that is normally done for executable into two parts. The `__managed__Startup` function runs classlib and module initializers. The `__managed__Main` function performs the remaining initialization and executes the entry-point of the managed application.

The use case for this is to allow calling `UnamanagedCallersOnly` functions before the managed `main` function. Running on iOS is the first use case for this feature (#80905).

It was not clear to me how to fit this into the larger NativeAot build system. Should this be thought of as "a static library that also has a `__managed__Main` compiled in" or as "an executable that splits its initialization into two parts"? So I added support to ILC for both cases: compiling with the `--nativelib` flag and without it.

Lastly, I added some build integration the "an executable that splits its initialization into two parts" case, along with test. The idea is the user sets the `CustomNativeMain` property in their project. They are then responsible for providing a `NativeLibrary` item pointing to an object file containing their native `main` function. At runtime, when they call their first managed function, NativeAOT initialization will run. This includes calling `__managed__Main`, so the user cannot forget to initialize the runtime.

Related issues: #81097 #77957

AustinWise committed:
[NativeAot] Add ability to generate library and exe entry points

AustinWise committed:
Add build integration and test for using split entry point

ivanpovazan commented on src/tests/nativeaot/CustomMain/CustomMainNative.cpp:
```diff
@@ -0,0 +1,22 @@
+#endif
+{
+    puts("hello from native code");
+    SetExitCodeInManagedSide(100);
```
Maybe I did not understand the intention correctly, but shouldn't `NativeAOT_StaticInitialization` be called prior to any managed side code.

AustinWise replied:
See the changes to the MSBuild files. They force the `NativeAOT_StaticInitialization` symbol to be defined. This causes the constructor of the `InitializeRuntimePointerHelper` class to run before the native main function. This registers a callback with the native AOT runtime. On the first reverse P/Invoke into managed code, the NativeAOT runtime will run the callback to initialize the runtime.

The `__managed_Main` function is also a reverse P/Invoke, so initialization can happen when it is called too.

ivanpovazan replied:
Ah my bad, I missed that part. Thanks for pointing it out.

AustinWise committed:
Add missing copyright headers.

MichalStrehovsky commented on src/tests/nativeaot/CustomMain/CustomMain.cs:
```diff
@@ -0,0 +1,25 @@
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+class Program
```
Could you also add a [module initializer](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/module-initializers) that increments some static int variable and then in Main check that the static variable is 1? (To make sure it executed, and it executed once)

AustinWise replied:
I added a module initializer that increments a static variable. Along with the class constructor, this should ensure that everything is executed once.

MichalStrehovsky commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -208,6 +208,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <IlcArg Include="@(RdXmlFile->'--rdxml:%(FullPath)')" />
       <IlcArg Include="@(TrimmerRootDescriptor->'--descriptor:%(FullPath)')" />
       <IlcArg Condition="'$(NativeLib)' != ''" Include="--nativelib" />
+      <IlcArg Condition="'$(CustomNativeMain)' == 'true'" Include="--split-exe-initalization" />
```
I don't have a good suggestion, just an observation: none of the other command line parameters have separate words with dashes.

AustinWise replied:
I shortened the flag to `--splitinit`, which is more in line with the other flags.

MichalStrehovsky reviewed (approved):
Looks good otherwise, thanks!

AustinWise committed:
Use ModuleInitializer rather than class constructor in test

A coding error is more likely to run the Module Initializer twice
than the cctor.

AustinWise committed:
Shorten flag name to --splitinit

And fix some spelling in the variable name.

AustinWise committed:
Add back the static constructor

And change values that are added up so they are not multiples of each other

MichalStrehovsky commented:
The **(Build CoreCLR Tools Unit Tests linux x64 checked)** failure is real. Otherwise looks good to merge, thanks!

```
src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/ILCompilerDriver.cs(65,30): error CS7036: (NETCORE_ENGINEERING_TELEMETRY=Build) There is no argument given that corresponds to the required parameter 'generateLibraryAndModuleInitializers' of 'MainMethodRootProvider.MainMethodRootProvider(EcmaModule, IReadOnlyCollection<MethodDesc>, bool)'
```

ivanpovazan commented:
LGTM as well, thank you @AustinWise very much!

Let me know if you need help fixing the remaining test failure.
Once we get that fixed, we can merge this.

As a follow-up, we should also update the documentation here: https://github.com/dotnet/samples/blob/main/core/nativeaot/NativeLibrary/README.md

AustinWise committed:
Fix test build.

AustinWise commented:
Opps, I missed that test. It should be fixed now.

> As a follow-up, we should also update the documentation here: https://github.com/dotnet/samples/blob/main/core/nativeaot/NativeLibrary/README.md

Sure, I can follow up with a sample based on the unit test.
